### PR TITLE
Handle canceled requests

### PR DESCRIPTION
**Pull request by v-mserdar**
## Description
A 500 response was occurring for cases where a request was canceled. The underlying exception was a System.OperationCanceledException.

## Related issues
Addresses [#86094](https://microsofthealth.visualstudio.com/Health/_workitems/edit/86094).

## Testing
I tried to set up a test in the FhirStorageTest.cs as well as ExceptionTests.cs files but had difficulty in doing so as I wasn't sure on how to introduce a CancellationTokenSource that gets canceled.

Having said that, I tested locally by doing the following.

**For Comosdb:** In the CosmosQuery.cs file I temporarily added the following code prior to line 77 so that the token that gets passed to _feedIterator.ReadNextAsync(token) is in a canceled state.
      using var ctx = new CancellationTokenSource();
      ctx.Cancel();
      token = ctx.Token;

**For SQL:** I temporarily revised SearchService.cs by adding the following at line 50 so that it also sends a canceled token to SearchAsync.
      if (resourceType.Equals("Patient", StringComparison.OrdinalIgnoreCase))
      {
          using var ctx = new CancellationTokenSource();
          ctx.Cancel();
          cancellationToken = ctx.Token;
      }

With both of the temporary revisions, I used the .http files for sending a request like so to my locally running instance.
GET {{hostname}}/Patient?gender:not=female

**SQL db returns this:**

HTTP/1.1 408 Request Timeout
Connection: close
Content-Type: application/fhir+json; charset=utf-8
Date: Thu, 20 Oct 2022 16:36:23 GMT
Server: Kestrel
Transfer-Encoding: chunked
X-Request-Id: 032b5f18-3890-4057-a7f1-9aa33dfd5466
X-Instance-Id: ----254e919c-e5f1-41be-8ecb-30bd4157905a
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Content-Security-Policy: frame-src 'self';

{
  "resourceType": "OperationOutcome",
  "id": "032b5f18-3890-4057-a7f1-9aa33dfd5466",
  "meta": {
    "lastUpdated": "2022-10-20T16:36:24.0008425+00:00"
  },
  "issue": [
    {
      "severity": "error",
      "code": "timeout",
      "diagnostics": "A task was canceled."
    }
  ]
}

**CosmosDb returns this:**

HTTP/1.1 408 Request Timeout
Connection: close
Content-Type: application/fhir+json; charset=utf-8
Date: Thu, 20 Oct 2022 16:41:44 GMT
Server: Kestrel
Transfer-Encoding: chunked
X-Request-Id: 13de7263-4231-4296-aae1-84d3fd1c4d8c
X-Instance-Id: ----b21eb847-0d3b-42fe-a6c4-0635d9150c95
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Content-Security-Policy: frame-src 'self';

{
  "resourceType": "OperationOutcome",
  "id": "13de7263-4231-4296-aae1-84d3fd1c4d8c",
  "meta": {
    "lastUpdated": "2022-10-20T16:41:45.4974551+00:00"
  },
  "issue": [
    {
      "severity": "error",
      "code": "timeout",
      "diagnostics": "The operation was canceled.\r\nCancellation Token has expired: True. Learn more at: https://aka.ms/cosmosdb-tsg-request-timeout\r\nCosmosDiagnostics: {\"Summary\":{},\"name\":\"Typed FeedIterator ReadNextAsync\",\"id\":\"0c2c3aef-514f-4133-92bd-1ba569307d0e\",\"start time\":\"04:41:43:113\",\"duration in milliseconds\":5.0934,\"data\":{\"Client Configuration\":{\"Client Created Time Utc\":\"2022-10-20T16:40:06.5747707Z\",\"MachineId\":\"hashedMachineName:bba927d5-5fb5-26cb-61dd-74d0d5c0c617\",\"NumberOfClientsCreated\":1,\"NumberOfActiveClients\":1,\"ConnectionMode\":\"Direct\",\"User Agent\":\"cosmos-netstandard-sdk/3.31.0|1|X64|Microsoft Windows 10.0.19043|.NET 6.0.9|N|\",\"ConnectionConfig\":{\"gw\":\"(cps:50, urto:10, p:False, httpf: False)\",\"rntbd\":\"(cto: 5, icto: -1, mrpc: 30, mcpe: 65535, erd: True, pr: ReuseUnicastPort)\",\"other\":\"(ed:False, be:False)\"},\"ConsistencyConfig\":\"(consistency: Session, prgns:[], apprgn: )\",\"ProcessorCount\":12},\"Query Correlated ActivityId\":\"d177ce76-b4ee-432d-8fb3-c661eebd4516\"},\"children\":[{\"name\":\"Create Query Pipeline\",\"id\":\"f5260451-bbd3-4987-a4ad-dc414b742373\",\"start time\":\"04:41:43:113\",\"duration in milliseconds\":1.3414,\"children\":[{\"name\":\"Get Container Properties\",\"id\":\"7b19b4c4-7f3e-478d-93c9-7510cb1dc41d\",\"start time\":\"04:41:43:113\",\"duration in milliseconds\":0.6123,\"children\":[{\"name\":\"Get Collection Cache\",\"id\":\"a11af642-a3d9-4872-a38d-ed055ee06a06\",\"start time\":\"04:41:43:113\",\"duration in milliseconds\":0.0012}]}]},{\"name\":\"CosmosOperationCanceledException\",\"id\":\"22f0b63f-95eb-45fd-bbde-c91b63e02b00\",\"start time\":\"04:41:43:116\",\"duration in milliseconds\":0.0071,\"data\":{\"Operation Cancelled Exception\":\"System.OperationCanceledException: The operation was canceled.\\r\\n   at System.Threading.CancellationToken.ThrowOperationCanceledException()\\r\\n   at Microsoft.Azure.Cosmos.Common.CollectionCache.ResolveByNameAsync(String apiVersion, String resourceAddress, Boolean forceRefesh, ITrace trace, IClientSideRequestStatistics clientSideRequestStatistics, CancellationToken cancellationToken)\\r\\n   at Microsoft.Azure.Cosmos.ClientContextCore.GetCachedContainerPropertiesAsync(String containerUri, ITrace trace, CancellationToken cancellationToken)\\r\\n   at Microsoft.Azure.Cosmos.CosmosQueryClientCore.GetCachedContainerQueryPropertiesAsync(String containerLink, Nullable`1 partitionKey, ITrace trace, CancellationToken cancellationToken)\\r\\n   at Microsoft.Azure.Cosmos.Query.Core.ExecutionContext.CosmosQueryExecutionContextFactory.TryCreateCoreContextAsync(DocumentContainer documentContainer, CosmosQueryContext cosmosQueryContext, InputParameters inputParameters, ITrace trace, CancellationToken cancellationToken)\\r\\n   at Microsoft.Azure.Cosmos.Query.Core.AsyncLazy`1.GetValueAsync(ITrace trace, CancellationToken cancellationToken)\\r\\n   at Microsoft.Azure.Cosmos.Query.Core.Pipeline.LazyQueryPipelineStage.MoveNextAsync(ITrace trace)\\r\\n   at Microsoft.Azure.Cosmos.Query.Core.Pipeline.NameCacheStaleRetryQueryPipelineStage.MoveNextAsync(ITrace trace)\\r\\n   at Microsoft.Azure.Cosmos.Query.Core.Pipeline.CatchAllQueryPipelineStage.MoveNextAsync(ITrace trace)\\r\\n   at Microsoft.Azure.Cosmos.Query.QueryIterator.ReadNextAsync(ITrace trace, CancellationToken cancellationToken)\"}}]}"
    }
  ]
}

## FHIR Team Checklist
- [X] **Update the title** of the PR to be succinct and less than 50 characters
- [X] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [X] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [X] Tag the PR with **Azure API for FHIR** if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [ ] Tag the PR with **Azure Healthcare APIs** if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

**Comment by rbans96** [comment]
Could we add a unit test for this?

**Comment by v-mserdar** [comment]
> 

This is the part where I had difficulty. Can you recommend a location in our code base where this should be? Also, how to trigger the cancel from a web endpoint?

**Review by PTaladay — APPROVED** [review]

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Filters/OperationOutcomeExceptionFilterAttribute.cs [hunk 1]
```diff
@@ -227,6 +227,11 @@ public override void OnActionExecuted(ActionExecutedContext context)
                 context.Result = CreateOperationOutcomeResult(formatException.Message, OperationOutcome.IssueSeverity.Error, OperationOutcome.IssueType.Invalid, HttpStatusCode.BadRequest);
                 context.ExceptionHandled = true;
             }
+            else if (context.Exception is System.OperationCanceledException operationCanceledException)
+            {
+                context.Result = CreateOperationOutcomeResult(operationCanceledException.Message, OperationOutcome.IssueSeverity.Error, OperationOutcome.IssueType.Timeout, HttpStatusCode.RequestTimeout);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Filters/OperationOutcomeExceptionFilterAttribute.cs:232]: We should return a more generic message, not these details

Comment (v-mserdar) [reply src/Microsoft.Health.Fhir.Shared.Api/Features/Filters/OperationOutcomeExceptionFilterAttribute.cs:232]: @brendankowitz Thanks for the review. Would you mind giving some guidance on what the generic message should be? Would a message such as "An operation was canceled" suffice?

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Api/Features/Filters/OperationOutcomeExceptionFilterAttribute.cs:232]: Adding @EXPEkesheth for thoughts.

I think we should have something along the lines of `The attempted operation was cancelled, please retry the request. If you believe you have received this message in error, contact Customer Support.`

**Review by brendankowitz — APPROVED** [review]

**Pull request merged**